### PR TITLE
Fix multiclass MCC loss function

### PR DESCRIPTION
**Pull request by kevingreenman**
## Description
Fix our implementation of multiclass Matthews correlation coefficient (MCC) loss function to give results that match sklearn and add unit tests for this.

## Relevant issues
This addresses [Issue 306](https://github.com/chemprop/chemprop/issues/306). I originally thought that the only problem was with the loss being infinite when all predictions and/or all targets were the same. My original plan was to raise a `ValueError` in this case and recommend that the user choose a different loss function. However, while writing unit tests for this, I realized that our existing implementation of MCC did not give the same results as [sklearn's implementation](https://scikit-learn.org/stable/modules/model_evaluation.html#matthews-corrcoef), even for normal cases. I also saw that sklearn's [convention](https://github.com/scikit-learn/scikit-learn/blob/8175cd60e3df606905aedf2dd3ba0832e50cd963/sklearn/metrics/_classification.py#L917) is to return 0 incases where the loss is infinite. I fixed the problem with our implementation and updated the code to return 0 in cases of infinite loss rather than raise an error message that might make this function unusable for many datasets (especially imbalanced datasets).

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

**Review by davidegraff — COMMENTED** [review]
I think I have some comments on code that's not yours but I recall that I've already made them on another PR... ignore them for now!

**Comment by kevingreenman** [comment]
This is already based on master, so maybe these are in another open PR?

File: chemprop/train/loss_functions.py [hunk 1]
```diff
@@ -127,21 +123,31 @@ def mcc_multiclass_loss(
...
+
+    t_sum = torch.sum(bin_targets * data_weights * mask, axis=0)  # number of times each class truly occurred
+    p_sum = torch.sum(bin_preds * data_weights * mask, axis=0)  # number of times each class was predicted
+
+    n_correct = torch.sum(bin_preds * bin_targets * data_weights * mask)  # total number of samples correctly predicted
+    n_samples = torch.sum(predictions * data_weights * mask)  # total number of samples
+
+    cov_ytyp = n_correct * n_samples - torch.dot(p_sum, t_sum)
+    cov_ypyp = n_samples**2 - torch.dot(p_sum, p_sum)
+    cov_ytyt = n_samples**2 - torch.dot(t_sum, t_sum)
```
[the remark this answers is not in the document]
Comment (kevingreenman) [reply chemprop/train/loss_functions.py:144]: I think they're about the same readability in general, but `torch.dot` aligns with the `np.dot` in sklearn's implementation of MCC, so I'll leave as is in case anyone compares them

**Comment by kevingreenman** [comment]
@davidegraff thanks for the suggestions! I think I've addressed all of them that are related to my changes

**Comment by kevingreenman** [comment]
The failed tests were a result of replacing `torch.arange(predictions.shape[0])` with `:len(predictions)`. I didn't realize at first that these would have different behavior due to the shape of `predictions`. I've reverted this change, and all tests pass on my local now.

File: chemprop/train/loss_functions.py [hunk 2]
```diff
@@ -133,18 +133,20 @@ def mcc_multiclass_loss(
...
+    n_correct = torch.sum(bin_preds * bin_targets * masked_data_weights)  # total number of samples correctly predicted
+    n_samples = torch.sum(predictions * masked_data_weights)  # total number of samples
 
     cov_ytyp = n_correct * n_samples - torch.dot(p_sum, t_sum)
     cov_ypyp = n_samples**2 - torch.dot(p_sum, p_sum)
     cov_ytyt = n_samples**2 - torch.dot(t_sum, t_sum)
 
     if cov_ypyp * cov_ytyt == 0:
-        loss = torch.tensor(0.0, dtype=torch.float64)
+        loss = torch.tensor(0.0, dtype=torch.float)
```
Comment (davidegraff) [inline chemprop/train/loss_functions.py:149]: is is necessary to specify?
```python
>>> x = torch.tensor(0.0)
>>> x.type
torch.float32
>>> y = torch.tensor(0.0, dtype=torch.float)
>>> y.dtype
torch.float32
```

**Review by davidegraff — APPROVED** [review]
think this is is my last comment- otherwise looks great! thanks!

**Comment by davidegraff** [comment]
what's the shape of `predictions`?
